Feat/add updateAcr option to approveAccessRequest function

solid-akb opened the pull request:
<!-- When adding a new feature: -->

# Add updateAcr option to the approveAccessRequest function

# Checklist

- [x] All acceptance criteria are met.
- [x] Relevant documentation, if any, has been written/updated.
- [x] The changelog has been updated, if applicable.
- [x] New functions/types have been exported in `index.ts`, if applicable.
- [x] New modules (i.e. new `.ts` files) are listed in the `exports` field in `package.json`, if applicable.
- [x] New modules (i.e. new `.ts` files) are listed in the `typedocOptions.entryPoints` field in `tsconfig.json`, if applicable.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

solid-akb committed:
flag added and true tests work

solid-akb committed:
flag to false works

solid-akb committed:
added docs and Changelog

solid-akb committed:
tests passing

solid-akb committed:
updated comments

solid-akb committed:
moved SaveInContainerOptions to type directory

NSeydoux commented on src/resource/saveSolidDatasetInContainer.ts:
```diff
@@ -27,18 +27,14 @@ import {
- * using [[createContainerAt]], or directly save the SolidDataset at the desired location using
- * [[saveSolidDatasetAt]].
+ * using createContainerAt, or directly save the SolidDataset at the desired location using
+ * {@link resource.saveSolidDatasetAt}.
```
Good catch!

ThisIsMissEm replied:
Huh! Is that how you do links within the one piece of documentation in typedoc? Neat.

NSeydoux committed:
Update TSDoc comment

NSeydoux committed:
Update the changelog

NSeydoux committed:
lint

NSeydoux committed:
Remove redundant test portion

NSeydoux committed:
Fix test typings

NSeydoux committed:
Make updatedAcr flag boolean

NSeydoux committed:
Use nullish coalescing

NSeydoux commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
 
 - Added the `createContainerInContainer` method.
 - Added `saveSolidDatasetInContainer` method.
+- Added the `updateAcr` to the list of options that can be included when calling the `approveAccessRequest` method. `updateAcr` will default to `true` if nothing is passed in.
```
8d50ecbd6bf11eca020a9491a8fb167a40099d0e

NSeydoux commented on e2e/node/e2e.test.ts:
```diff
@@ -79,7 +79,7 @@ describe(`End-to-end access grant tests for environment [${environment}}]`, () =
 
   // Setup the shared file
-  beforeAll(async () => {
+  beforeEach(async () => {
```
368e457861962abbe7c24efe620378d69ab22b2e

NSeydoux commented on src/gConsent/type/AccessBaseOptions.ts:
```diff
@@ -46,6 +46,7 @@ import type { UrlString } from "@inrupt/solid-client";
  * @since 0.4.0
  */
 export interface AccessBaseOptions {
+  updateAcr?: any;
```
0b46b823b26c9efa420cd5098402caf257e95707

chelseapinka reviewed (approved):
Looks good, backwards compatibility steps make sense.

ThisIsMissEm commented on src/gConsent/manage/approveAccessRequest.test.ts:
```diff
@@ -24,7 +24,7 @@ import { jest, it, describe, expect } from "@jest/globals";
 import type { issueVerifiableCredential } from "@inrupt/solid-client-vc";
 
-import type * as SolidClient from "@inrupt/solid-client";
+import * as SolidClient from "@inrupt/solid-client";
```
I think you're only importing types here, aren't you?

ThisIsMissEm commented on src/gConsent/manage/approveAccessRequest.test.ts:
```diff
@@ -64,47 +64,61 @@ jest.mock("cross-fetch");
   return solidClientModule;
 };
 
+const mockedInitialResource = mockSolidDatasetFrom(
```
@NSeydoux what do you think about making `mockSolidDatasetFrom` accept a generic? So:

```
const mockedInitialResource = mockSolidDatasetFrom<
  SolidClient.WithServerResourceInfo & SolidClient.WithAcp
>(
"https://some.resource"
);
```

This might be better than casting to unknown

NSeydoux replied:
I would prefer not to: this cast is quite hacky, and it's only okay because it just ends up satisfying the mock defined here: 
```
solidClientModule.acp_ess_2.getResourceInfoWithAcr.mockResolvedValueOnce(
      options?.initialResource
    );
``` 
Arguably, it would be possible to add the cast to unknown there, rather than when initialling the variable. In any case, that's a dirty cast, and I'd rather keep the mock object have an accurate type, rather than picking it up from the generic, since `mockSolidDatasetFrom` is part of the public API, so I wouldn't want it to allow to do things that are too crazy.

ThisIsMissEm replied:
hmm, yeah, let's do the casting at the call site then, I guess? idk. Minor issue tbh.

ThisIsMissEm replied:
Maybe we'd more want something like `mockSolidDatasetFrom(resourceIri: string, { withAcr: string / ACR object })` or something?

ThisIsMissEm commented on src/gConsent/manage/approveAccessRequest.ts:
```diff
@@ -140,23 +141,33 @@ async function internal_approveAccessRequest(
-    internalOptions
-  );
+
+  if (internalOptions.updateAcr) {
```
perhaps?
```suggestion
  if (internalOptions.updateAcr === true) {
```

ThisIsMissEm commented on src/gConsent/type/AccessBaseOptions.ts:
```diff
@@ -46,6 +46,7 @@ import type { UrlString } from "@inrupt/solid-client";
  * @since 0.4.0
  */
 export interface AccessBaseOptions {
+  updateAcr?: boolean;
```
```suggestion
  /**
   * @since unreleased
   */
  updateAcr?: boolean;
```

ThisIsMissEm commented on src/resource/createContainerInContainer.ts:
```diff
@@ -46,17 +43,14 @@ interface SaveInContainerOptions extends FetchOptions {
- * function to make the HTTP request, compatible with the browser-native [fetch
- * API](https://developer.mozilla.org/docs/Web/API/WindowOrWorkerGlobalScope/fetch#parameters).
- * `options.slugSuggestion` accepts a string for your new Container's name.
+ * @param options Optional parameter `options.fetch`: An alternative `fetch` function to make the HTTP request, compatible with the browser-native [fetch API](https://developer.mozilla.org/docs/Web/API/WindowOrWorkerGlobalScope/fetch#parameters).
```
It should be complaining about it.. I wonder if we've contention between prettier and eslint here maybe?

NSeydoux replied:
Hum, I wonder if this would explain it?: https://prettier.io/docs/en/options.html#prose-wrap? I'll manually break that down for now.

ThisIsMissEm replied:
It wouldn't be that because this repo uses Prettier 2.3.2 (we should absolutely upgrade though, lol)

ThisIsMissEm commented on src/type/SaveInContainerOptions.ts:
```diff
@@ -0,0 +1,38 @@
+ *
+ * @since unreleased
+ */
+export interface SaveInContainerOptions extends FetchOptions {
```
Can we keep this TODO:
```suggestion
// TODO: This should come from @inrupt/solid-client
export interface SaveInContainerOptions extends FetchOptions {
```

ThisIsMissEm reviewed (commented):
Looks good, a few suggested changes

NSeydoux committed:
Update CHANGELOG.md

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux commented on e2e/node/e2e.test.ts:
```diff
@@ -251,6 +243,64 @@ describe(`End-to-end access grant tests for environment [${environment}}]`, () =
       ]);
     });
+
+    it("will issue an access request, grant access to a resource, but will not update the ACR if the updateAcr flag is set to false", async () => {
```
I'd prefer to avoid spying on calls for the end-to-end tests, and keeping the tests focused on tangible external elements. in this case, checking the value of the ACR after the call shows that it hasn't been updated as we expect, and it doesn't make any assumptions on how that update may have happened.

I'll add an e2e test for when the flag is set to true.

ThisIsMissEm replied:
Hmm, okay! That's fine then, just as long as we have test coverage asserting that the call is definitely made somewhere.

NSeydoux replied:
https://github.com/inrupt/solid-client-access-grants-js/pull/432/commits/7ff815ac24a15f7acb69549e3054e3d4e30c37a7

NSeydoux committed:
Update e2e/node/e2e.test.ts

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Explicitly import types

NSeydoux committed:
Update src/gConsent/manage/approveAccessRequest.ts

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Update src/gConsent/type/AccessBaseOptions.ts

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
lint

NSeydoux committed:
Update src/type/SaveInContainerOptions.ts

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Add e2e test checking for ACR update

NSeydoux committed:
Fix e2e test typo

chelseapinka reviewed (approved):
Looks good :+1: except some testing issues

NSeydoux committed:
Extent e2e test OpenID timeout

Switching to logging in each test instead of one for the whole suite creates some lags, so this increases the timeout.

NSeydoux commented:
It also looks like the node e2e tests have been made flakier by a change from authenticating once per test suite to authenticating once per test. https://github.com/inrupt/solid-client-access-grants-js/pull/432/commits/7d282d984b5647c917703f786ef3f58c433ab40e should make things more reliable.
